### PR TITLE
11416 MethodType.toString prints in scala format

### PR DESCRIPTION
**Pull request by psilospore**
Issue: https://github.com/scala/bug/issues/11416
I skimmed through the checkfiles and it seems right to me

### CODE REVIEW

File: test/files/neg/names-defaults-neg.check [hunk 1]
```diff
@@ -45,26 +45,26 @@ names-defaults-neg.scala:28: error: Int does not take parameters
...
-and  method f in object t1 of type (a: Int, b: String)String
+both method f in object t1 of type (b: String, a: Int): String
+and  method f in object t1 of type (a: Int, b: String): String
 match argument types (b: String,a: Int)
   t1.f(b = "dkljf", a = 1)
      ^
 names-defaults-neg.scala:44: error: ambiguous reference to overloaded definition,
-both method f in object t3 of type (a2: Int)(b: Int)String
-and  method f in object t3 of type (a1: Int)String
+both method f in object t3 of type (a2: Int): (b: Int): String
```
Comment (azolotko) [inline test/files/neg/names-defaults-neg.check]: Does this format of multiple parameter lists look as intended? `(a2: Int): (b: Int): String` looks as strange as `(a2: Int)(b: Int)String` did, isn't it?

File: src/reflect/scala/reflect/internal/Types.scala [hunk 2]
```diff
@@ -2913,7 +2913,7 @@ trait Types
     //TODO this may be generalised so that the only constraint is dependencies are acyclic
     def approximate: MethodType = MethodType(params, resultApprox)
 
-    override def safeToString = paramString(this) + resultType
+    override def safeToString = s"${paramString(this)}: $resultType"
```
Comment (hrhino) [inline src/reflect/scala/reflect/internal/Types.scala]: I think you need to check if `resultType` is a `MethodType` and not add the colon if so. Maybe look at the "object creation impossible" error in `RefChecks` for how it's done there.

File: test/files/neg/macro-invalidsig.check [hunk 3]
```diff
@@ -73,13 +73,13 @@ Macros_Test_2.scala:46: error: type arguments [U] do not conform to method foo's
...
                                ^
-Macros_Test_2.scala:59: error: macro implementation reference has too few type arguments for method foo: [T, U, V](c: scala.reflect.macros.blackbox.Context)(implicit evidence$5: c.WeakTypeTag[T], implicit evidence$6: c.WeakTypeTag[U], implicit V: c.WeakTypeTag[V])c.Expr[Unit]
+Macros_Test_2.scala:59: error: macro implementation reference has too few type arguments for method foo: [T, U, V](c: scala.reflect.macros.blackbox.Context): (implicit evidence$5: c.WeakTypeTag[T], implicit evidence$6: c.WeakTypeTag[U], implicit V: c.WeakTypeTag[V]): c.Expr[Unit]
     def foo15[V]: Unit = macro Impls15.foo
                                        ^
-Macros_Test_2.scala:60: error: wrong number of type parameters for method foo: [T, U, V](c: scala.reflect.macros.blackbox.Context)(implicit evidence$7: c.WeakTypeTag[T], implicit evidence$8: c.WeakTypeTag[U], implicit V: c.WeakTypeTag[V])c.Expr[Unit]
+Macros_Test_2.scala:60: error: wrong number of type parameters for method foo: [T, U, V](c: scala.reflect.macros.blackbox.Context): (implicit evidence$7: c.WeakTypeTag[T], implicit evidence$8: c.WeakTypeTag[U], implicit V: c.WeakTypeTag[V]): c.Expr[Unit]
```
Comment (hrhino) [inline test/files/neg/macro-invalidsig.check]: There's a ticket somewhere adjacent to the fact that we print each param in an implicit param list with `implicit` before it. That enhancement would be in the spirit of this PR, I think.

`def isImplicitParams = params.nonEmpty && params.head.isImplicit` and so on.

Sorry, I mean `!params.isEmpty`. Is that old yet?

**Review by hrhino — CHANGES_REQUESTED** [review]

**Comment by SethTisue** [comment]
@psilospore interested in returning to this? PR queue is moving again, post-2.13.0

**Comment by psilospore** [comment]
Yeap @SethTisue I can start looking at this again
Edit: just as an update, I've gotten busy with other life events. I would still like to work on this but it may take time.

**Comment by psilospore** [comment]
The recent changes should also only have implicit once instead of before each parameter. e.g. `(a: A)(b: B)(implicit c: C, d: D): E`. I think that's what @hrhino was suggesting to include. Correct me if I'm wrong.

**Comment by lrytz** [comment]
This looks good!

Are we fine with the representation for constructors?

```diff
- t1038.scala:4: error: not enough arguments for constructor X: (x: Int)X.
+ t1038.scala:4: error: not enough arguments for constructor X: (x: Int): X.
```

We should probably also change the `toString` of `NullaryMethodType`:

```scala
scala> class C[-T] { def f: T = ??? }
                         ^
       error: contravariant type T occurs in covariant position in type => T of method f
```

**Comment by psilospore** [comment]
Thanks! What should the error message look like for NullaryMethodType? Would it be something like 
error: contravariant type T occurs in covariant position in type T of method f

**Comment by lrytz** [comment]
>  error: contravariant type T occurs in covariant position in type T of method f

I think that looks good.

**Comment by SethTisue** [comment]
@psilospore interested in returning this? needs a rebase, which will probably be easy. this could make 2.13.2, which we plan to release later this month

**Comment by psilospore** [comment]
@SethTisue thanks I've updated it!

**Comment by SethTisue** [comment]
```
!!  1 - res/t687.res                              [output differs][duration 5.97s]
!!  2 - res/t9170.res                             [output differs]
```

**Comment by psilospore** [comment]
@SethTisue sorry missed that it should be good now

**Comment by SethTisue** [comment]
we'd better merge this before any more merge conflicts accumulate :-)

**Comment by lrytz** [comment]
🎉🎉

**Pull request merged**